File: dotnet/source/amp.bus.rabbit/topology/Exchange.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace amp.bus.rabbit.topology
{
    public struct Exchange : IEquatable<Exchange>
    {
        public string Name { get; private set; }
        public string HostName { get; private set; }
        public string VirtualHost { get; private set; }
        public int Port { get; private set; }
        public string RoutingKey { get; private set; }
        public string QueueName { get; private set; }
        public string ExchangeType { get; private set; }
        public bool IsDurable { get; private set; }
        public bool IsAutoDelete { get; private set; }
        public IDictionary Arguments { get; private set; }


        public Exchange(
            string name, 
            string hostName, 
            string vHost, 
            int port, 
            string routingKey, 
            string queueName = null, 
            string exchangeType = "topic", 
            bool isDurable = false, 
            bool autoDelete = false, 
            IDictionary arguments = null)
            : this()
        {
            this.Name = name;
            this.HostName = hostName;
            this.VirtualHost = vHost;
            this.Port = port;
            this.RoutingKey = routingKey;
            this.QueueName = queueName;
            this.ExchangeType = exchangeType;
            this.IsDurable = isDurable;
            this.IsAutoDelete = autoDelete;
            this.Arguments = arguments;
        }


        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();

            sb.Append("{");
            sb.AppendFormat("Name:{0},", this.Name);
            sb.AppendFormat("HostName:{0},", this.HostName);
            sb.AppendFormat("VirtualHost:{0},", this.VirtualHost);
            sb.AppendFormat("Port:{0},", this.Port);
            sb.AppendFormat("RoutingKey:{0},", this.RoutingKey);
            sb.AppendFormat("QueueName:{0},", this.QueueName);
            sb.AppendFormat("ExchangeType:{0},", this.ExchangeType);
            sb.AppendFormat("IsDurable:{0},", this.IsDurable);
            sb.AppendFormat("IsAutoDelete:{0},", this.IsAutoDelete);

            sb.Append("}");
            return sb.ToString();
        }


        public override int GetHashCode()
        {
            unchecked
            {
                int hash = 71;

                hash = hash * 97 + this.Name.GetHashCode();
                hash = hash * 97 + this.HostName.GetHashCode();
                hash = hash * 97 + this.VirtualHost.GetHashCode();
                hash = hash * 97 + this.Port.GetHashCode();

                return hash;
            }
        }

        public override bool Equals(object obj)
        {
            bool isEqual = false;

            try
            {
                if (obj is Exchange)
                {
                    isEqual = this.Equals((Exchange)obj);
                }
            }
            catch { }

            return isEqual;
        }
        
        public bool Equals(Exchange other)
        {
            return this.GetHashCode() == other.GetHashCode();
        }
    }
}

File: java/amp-rabbit/src/main/java/amp/rabbit/topology/Exchange.java
package amp.rabbit.topology;


import java.util.Map;

public class Exchange extends AmqpBaseModel {

	protected String exchangeType = "direct";
	
    public Exchange() {}
    
    /** 
     * ORIGINAL CONSTRUCTOR STRING:
     *              String name, String hostName, String vHost, int port, String routingKey, String queueName,
                    String exchangeType, boolean isDurable, boolean autoDelete, Map arguments
     * DELTA:
     * 	hostName DELETED,
     *  routingKey Deleted
     *  queueName Deleted
     *  
     * @param exchangeName
     * @param exchangeType
     * @param isAutoDelete
     * @param isDurable
     * @param shouldDeclare
     * @param virtualHost
     * @param arguments
     */
    public Exchange(String exchangeName, String exchangeType,
			boolean isAutoDelete, boolean isDurable, boolean shouldDeclare,
			Map<String, Object> arguments) {

		super(exchangeName, isAutoDelete, isDurable, shouldDeclare,
				arguments);
		this.exchangeType = exchangeType;
    }
    
    public String getExchangeType() {
    	return exchangeType;
    }    

    public void setExchangeType(String exchangeType) {
        this.exchangeType = exchangeType;
    }
 
    @Override
	public int hashCode() {
		final int prime = 31;
		int result = super.hashCode();
		result = prime * result
				+ ((exchangeType == null) ? 0 : exchangeType.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (!super.equals(obj))
			return false;
		//Not needed as super performs this test
		//if (getClass() != obj.getClass())
		//	return false;
		Exchange other = (Exchange) obj;
		if (exchangeType == null) {
			if (other.exchangeType != null)
				return false;
		} else if (!exchangeType.equals(other.exchangeType))
			return false;
		return true;
	}

	@Override
    public String toString() {
    	return 
    		super.getCommonStringInfo()
    			.add("ExchangeType",exchangeType)
    			.toString();
    }
    
    public static ExchangeBuilder builder(){
    	
		return new ExchangeBuilder();
}

public static class ExchangeBuilder {
	
		Exchange exchange = new Exchange();
	
		public ExchangeBuilder name(String exchangeName){
			
			this.exchange.setName(exchangeName);
			
			return this;
		}
		
	public ExchangeBuilder type(String exchangeType){
	    			
			this.exchange.setExchangeType(exchangeType);
			
			return this;
		}
	
	public ExchangeBuilder isDurable(boolean isDurable){
		
		this.exchange.setDurable(isDurable);
		
		return this;
	}
	
	public ExchangeBuilder isAutoDelete(boolean isAutoDelete){
		
		this.exchange.setAutoDelete(isAutoDelete);
		
		return this;
	}
	
	public ExchangeBuilder declare(boolean trueIfShouldDeclare){
		
		this.exchange.setShouldDeclare(trueIfShouldDeclare);
		
		return this;
	}
	
	public ExchangeBuilder arguments(Map<String, Object> arguments){
		
		if (arguments != null){
		
			this.exchange.setArguments(arguments);
		}
		return this;
	}
		
	public Exchange build(){
		
		return this.exchange;
	}
}

    
}
